src/http.zig

```diff
@@ -2368,12 +2417,37 @@ fn printResponse(response: picohttp.Response) void {
     Output.flush();
 }
 
+pub fn onPreconnect(this: *HTTPClient, comptime is_ssl: bool, socket: NewHTTPContext(is_ssl).HTTPSocket) void {
+    log("onPreconnect({})", .{this.url});
+    _ = socket_async_http_abort_tracker.swapRemove(this.async_http_id);
+    const ctx = if (comptime is_ssl) &http_thread.https_context else &http_thread.http_context;
+
+    ctx.releaseSocket(
+        socket,
+        this.did_have_handshaking_error and !this.reject_unauthorized,
+        this.url.hostname,
+        this.url.getPortAuto(),
+    );
+
+    this.state.reset(this.allocator);
+    this.state.response_stage = .done;
+    this.state.request_stage = .done;
+    this.state.stage = .done;
+    this.proxy_tunneling = false;
+    this.result_callback.run(@fieldParentPtr("client", this), HTTPClientResult{ .fail = null, .metadata = null, .has_more = false });
+}
+
 pub fn onWritable(this: *HTTPClient, comptime is_first_call: bool, comptime is_ssl: bool, socket: NewHTTPContext(is_ssl).HTTPSocket) void {
     if (this.signals.get(.aborted)) {
         this.closeAndAbort(is_ssl, socket);
         return;
     }
 
+    if (this.is_preconnect_only) {
+        this.onPreconnect(is_ssl, socket);
```

Comment:
should we call onPreconnect inside `onOpen` instead? we are sending a GET request or just connecting? What happens if the GET request have a big body? and whats happens if the server is not writable before we send the first request content? If is pre-flight should it be a HEAD request?
We need tests for this scenarios.
If we are doing a pre-flight we should wait until the request is complete (aka requirects, body etc), if we send another request when still receiving body when reusing the socket we will get a mal formed request and async pipelining is not really supported by a lot of servers.

Reply:
preconnect does not write any data. it just keeps the socket open. we don't know what the data they're sending will be, so we can't make any assumptions.